### PR TITLE
Added conflicting virtualservice host analyzer

### PR DESCRIPTION
**Pull request by nrjpoddar**
Fixes: #17504

Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[x] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: galley/pkg/config/analysis/analyzers/analyzers_test.go [hunk 1]
```diff
@@ -138,6 +138,15 @@ var testGrid = []testCase{
 			{msg.IstioProxyVersionMismatch, "Pod/enabled-namespace/details-v1-pod-old"},
 		},
 	},
+	{
+		name:       "virtualServiceConflictingHosts",
+		inputFiles: []string{"testdata/virtualservice_conflictinghosts.yaml"},
+		analyzer:   &virtualservice.ConflictingHostsAnalyzer{},
+		expected: []message{
+			{msg.ConflictingVirtualServiceHosts, ""},
+			{msg.ConflictingVirtualServiceHosts, ""},
```
Comment (nrjpoddar) [inline galley/pkg/config/analysis/analyzers/analyzers_test.go]: @sushicw the current way of verifying messages in unit tests is insufficient for this analyzer. We should verify the entire string message in my opinion.

Comment (nrjpoddar) [reply galley/pkg/config/analysis/analyzers/analyzers_test.go]: Maybe, I think having string comparison in tests forces us to follow good practices and make sure that the template string after substitution still makes sense for users.

I can add the optional third string.

File: galley/pkg/config/analysis/analyzers/testdata/virtualservice_conflictinghosts.yaml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        host: reviews
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: ratings
+  namespace: foo
+spec:
+  hosts:
+  - ratings # should generate an error as this conflicts with VirtualService bar/reviews
```
[the remark this answers is not in the document]
Comment (nrjpoddar) [reply galley/pkg/config/analysis/analyzers/testdata/virtualservice_conflictinghosts.yaml]: yes, updated.

File: galley/pkg/config/analysis/analyzers/util/hosts.go [hunk 3]
```diff
@@ -47,3 +52,16 @@ func getNamespaceAndNameFromFQDN(fqdn string) (string, string) {
...
+		return ""
+	} else if strings.HasPrefix(host, "*") {
+		return host
+	} else if strings.Contains(host, ".") {
+		return host
+	} else {
+		// need to return Fully Qualified Domain Name
+		return host + "." + namespace + "." + DefaultKubernetesDomain
+	}
+}
```
[the remark this answers is not in the document]
Comment (nrjpoddar) [reply galley/pkg/config/analysis/analyzers/util/hosts.go]: Good point, converting short names to FQDN for this analyzer seems more natural. Converting a FQDN to "namespace/name" requires regex matching which is doable but feels a bit weird.

File: galley/pkg/config/analysis/analyzers/virtualservice/conflictinghosts.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	}
+}
+
+// Analyze implements Analyzer
+func (c *ConflictingHostsAnalyzer) Analyze(ctx analysis.Context) {
+	hs := initSidecarHostsVirtualServices(ctx)
+	for h, vsNames := range hs {
+		if len(vsNames) > 1 {
+			ctx.Report(metadata.IstioNetworkingV1Alpha3Virtualservices,
+				msg.NewConflictingVirtualServiceHosts(nil, strings.Join(vsNames, ","), h))
```
[the remark this answers is not in the document]
Comment (nrjpoddar) [reply galley/pkg/config/analysis/analyzers/virtualservice/conflictinghosts.go]: That's a good point. I can update the logic to have reports for each conflicting VirtualService.

File: galley/pkg/config/analysis/analyzers/virtualservice/conflictinghosts.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	ctx.ForEach(metadata.IstioNetworkingV1Alpha3Virtualservices, func(r *resource.Entry) bool {
+		vs := r.Item.(*v1alpha3.VirtualService)
+		vsNs, _ := r.Metadata.Name.InterpretAsNamespaceAndName()
+		vsAttachedToSidecar := false
+		// No entry in gateways imply "mesh" by default
+		if len(vs.Gateways) == 0 {
+			vsAttachedToSidecar = true
+		} else {
+			for _, g := range vs.Gateways {
+				if g == meshGateway {
```
[the remark this answers is not in the document]
Comment (nrjpoddar) [reply galley/pkg/config/analysis/analyzers/virtualservice/conflictinghosts.go]: Will do.

File: galley/pkg/config/analysis/analyzers/util/hosts.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nrjpoddar) [reply galley/pkg/config/analysis/analyzers/util/hosts.go]: The current logic for `GetResourceNameFromHost` is insufficient for my use case. Example:

* VirtualService vs1 in namespace ns1 host google.com
* VirtualService vs2 in namespace ns2 host google.com

`GetResourceNameFromHost` will return "ns1/google.com" and "ns2/google.com" which if I use as keys in my map will get stored as two entries even though the underlying host "google.com" is the same.

File: galley/pkg/config/analysis/analyzers/util/hosts.go [hunk 6]
```diff
@@ -16,10 +16,15 @@ package util
...
 
 	"istio.io/istio/galley/pkg/config/resource"
 )
 
+const (
+	DefaultKubernetesDomain = "svc.cluster.local"
+)
+
 var (
 	fqdnPattern = regexp.MustCompile(`^(.+)\.(.+)\.svc\.cluster\.local$`)
```
[the remark this answers is not in the document]
Comment (nrjpoddar) [reply galley/pkg/config/analysis/analyzers/util/hosts.go]: it's an overkill to make that work, plus the regex is hard to parse.

**Comment by nrjpoddar** [comment]
@sushicw PTAL, I have addressed your review comments.

**Comment by nrjpoddar** [comment]
/test 
integ-istioctl-k8s-tests_istio

File: galley/pkg/config/analysis/msg/messages.yaml [hunk 7]
```diff
@@ -121,3 +121,13 @@ messages:
        - name: annotation
          type: string
 
+  - name: "ConflictingMeshGatewayVirtualServiceHosts"
+    code: IST0109
+    level: Error
+    description: "Conflicting hosts on VirtualServices associated with mesh gateway"
```
Comment (sushicw) [inline galley/pkg/config/analysis/msg/messages.yaml]: When you get a chance, please also see about creating some docs for this message (as described here: https://github.com/istio/istio/blob/master/galley/pkg/config/analysis/README.md#6-write-a-user-facing-documentation-page). 

We don't need to do that before shipping this PR, but let's do it before closing the related issue.

File: galley/pkg/config/analysis/analyzers/util/constants.go [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+const (
+	DefaultKubernetesDomain = "svc.cluster.local"
```
Comment (sushicw) [inline galley/pkg/config/analysis/analyzers/util/constants.go]: Should we move fqdnPattern (from hosts.go) into this file as well? Even if it's prohibitively painful to make the regex use the const, I think we're less likely to run into "forgot to update one of them" errors if we keep them adjacent.

File: galley/pkg/config/analysis/analyzers/virtualservice/conflictingmeshgatewayhosts.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+func combineResourceEntryNames(rList []*resource.Entry) string {
+	names := []string{}
+	for _, r := range rList {
+		names = append(names, r.Metadata.Name.String())
+	}
+	return strings.Join(names, ",")
+}
+
+func initSidecarHostsVirtualServices(ctx analysis.Context) map[string][]*resource.Entry {
```
Comment (sushicw) [inline galley/pkg/config/analysis/analyzers/virtualservice/conflictingmeshgatewayhosts.go]: s/Sidecar/MeshGateway/

File: galley/pkg/config/analysis/analyzers/util/hosts.go [hunk 10]
```diff
@@ -47,3 +48,19 @@ func getNamespaceAndNameFromFQDN(fqdn string) (string, string) {
 	}
 	return result[0][2], result[0][1]
 }
+
+// GetScopedFqdnHostname converts the passed host to FQDN if needed and applies the passed scope.
+func GetScopedFqdnHostname(scope, namespace, host string) string {
```
Comment (sushicw) [inline galley/pkg/config/analysis/analyzers/util/hosts.go]: Should we give "scoped FQDN" its own named type? e.g. 
```type scopedFqdn string``` 

I think it might make semantics easier to follow and even give us a little bit of compile-time type checking that we're passing around what we think we are.

**Comment by sushicw** [comment]
> Could this same approach work for hosts bound to gateway? It gets a bit trickier since you'll need to treat each gateway server as an independent instance, but the same approach below should work (you might be able to treat 'mesh' as an imaginary gateway and then you could use the same approach below).
> 
> Then that would catch a mistake where multiple example.com rules are bound to the same gateway. If this seems reasonable, it would be ideal to at least capture this with a TODO/issue since the approach used here could be reused.

Are you thinking of https://github.com/istio/istio/issues/17560? Feel free to add comment(s) to the issue capturing the thought. It does seem natural to try to re-use at least some of the same logic.

File: galley/pkg/config/analysis/msg/messages.yaml (hunk 7 above)
Comment (nrjpoddar) [reply galley/pkg/config/analysis/msg/messages.yaml]: I was going to do that, so thanks for the note.

File: galley/pkg/config/analysis/analyzers/util/hosts.go (hunk 10 above)
Comment (nrjpoddar) [reply galley/pkg/config/analysis/analyzers/util/hosts.go]: yep, good idea. Will do.

**Comment by nrjpoddar** [comment]
> Could this same approach work for hosts bound to gateway? It gets a bit trickier since you'll need to treat each gateway server as an independent instance, but the same approach below should work (you might be able to treat 'mesh' as an imaginary gateway and then you could use the same approach below).
> 
> Then that would catch a mistake where multiple example.com rules are bound to the same gateway. If this seems reasonable, it would be ideal to at least capture this with a TODO/issue since the approach used here could be reused.

yes, it does makes sense. I tried to create a general utility for creating scoped hosts and mesh gateways. I'm anticipating re-using it in multiple places.

**Comment by nrjpoddar** [comment]
@sushicw PR should be good to go now, if I can get a LGTM. Thanks!

**Comment by howardjohn** [comment]
virtualServiceConflictingMeshGatewayHosts is failing, I think from this PR? @nrjpoddar

**Comment by nrjpoddar** [comment]
@howardjohn yes, it looks like the test case relies on the output order of messages which is coming out of map and is not stable.

**Comment by nrjpoddar** [comment]
@howardjohn this new commit should fix the unit tests.

**Comment by sushicw** [comment]
We just recently pushed a change to sort messages consistently, which should prevent this problem in the future (even if apparently it was awkwardly timed with this particular PR). https://github.com/istio/istio/pull/17940

**Comment by howardjohn** [comment]
/retest

didn't seem to trigger it

**Comment by nrjpoddar** [comment]
FYI @howardjohn for this PR I had a weird ping pong situation going where one of the checks passed after retesting, goes into merge pool, fails again and then again passes after retesting but fails in merge pool. 

It wastes a quite a bit of time from dev and contribution point of view. I know you are working on improving some aspects of this so thought I would let you know.

**Comment by nrjpoddar** [comment]
@sushicw I think we should cherry-pick this into release-1.4. WDYT?
